packages/viem/scripts/chaingen.ts

```diff
@@ -0,0 +1,159 @@
+import toml from '@iarna/toml'
+import { Eta } from 'eta'
+import fs from 'fs'
+import path from 'path'
+import type { Address, PublicClient } from 'viem'
+import { createPublicClient, erc20Abi, http } from 'viem'
+import { mainnet, sepolia } from 'viem/chains'
+
+// Hardcoded. Can take a more elaborate approach if needed.
+const NETWORKS = ['mainnet', 'sepolia']
+const SUPERCHAIN_REGISTRY_PATH = path.join(
+  '..',
+  '..',
+  'lib',
+  'superchain-registry',
+)
+
+type ChainDefinition = {
+  chainName: string
+  exportName: string
+  chainId: number
+  sourceChainId: number
+  rpc: string
+  explorer: string
+  nativeCurrency: NativeCurrency
+  l1Addresses: Record<string, Address>
+}
+
+type NativeCurrency = { name: string; symbol: string; decimals: number }
+
+/** Utility functions **/
+
+function camelCase(str: string): string {
+  const parts = str.replace(/-/g, ' ').replace(/_/g, ' ').split(' ')
+  return (
+    parts[0].toLowerCase() +
+    parts
+      .slice(1)
+      .map((part) => part[0].toUpperCase() + part.substring(1))
+      .join('')
+  )
+}
+
+async function nativeCurrency(
+  client: PublicClient,
+  address: Address | undefined,
+): Promise<NativeCurrency> {
+  if (!address) {
+    return { name: 'Ether', symbol: 'ETH', decimals: 18 }
+  }
+
+  const [name, symbol, decimals] = await Promise.all([
+    client.readContract({ address, abi: erc20Abi, functionName: 'name' }),
+    client.readContract({ address, abi: erc20Abi, functionName: 'symbol' }),
+    client.readContract({ address, abi: erc20Abi, functionName: 'decimals' }),
+  ])
+
+  return { name, symbol, decimals }
+}
+
+/** Chain Generation **/
+
+async function main() {
+  console.log('Running chain generation...')
+  const eta = new Eta({
+    views: './scripts/templates',
+    debug: true,
+    autoTrim: [false, false],
+  })
+
+  const mainnetHttp = process.env.MAINNET_RPC_URL
+    ? [process.env.MAINNET_RPC_URL]
+    : mainnet.rpcUrls.default.http
+  const mainnetClient = createPublicClient({
+    chain: { ...mainnet, rpcUrls: { default: { http: mainnetHttp } } },
+    transport: http(),
+  })
+
+  const sepoliaHttp = process.env.SEPOLIA_RPC_URL
+    ? [process.env.SEPOLIA_RPC_URL]
+    : sepolia.rpcUrls.default.http
+  const sepoliaClient = createPublicClient({
+    chain: { ...sepolia, rpcUrls: { default: { http: sepoliaHttp } } },
+    transport: http(),
+  })
+
+  for (const network of NETWORKS) {
+    console.log(`Generating ${network}`)
+    const client = network === 'mainnet' ? mainnetClient : sepoliaClient
+
+    const configPath = path.join(
+      SUPERCHAIN_REGISTRY_PATH,
+      'superchain',
+      'configs',
+      network,
+    )
+    const entries = fs
+      .readdirSync(configPath)
+      .filter((entry) => !entry.includes('superchain'))
```

Comment:
just curious, why is this needed? might be good to add a comment here

Reply:
every network dir within the `config/<network>` has a `superchain.toml` file. This file has the L1 superchain config for that nework

def worth a comment

Reply:
It looks like they keep the same filepath structure within `config`. I'm going to leave it as it is right now but we can adjust later